blocks/datasets/__init__.py

```diff
@@ -529,11 +541,66 @@ def get_data(self, request):
             self.cache[i] = cache[request:]
         return tuple(data)
 
+    def get_epoch_iterator(self, **kwargs):
+        self.cache = [[] for source in self.sources]
+        return super(CachedDataStream, self).get_epoch_iterator(**kwargs)
+
     def _cache(self):
         for cache, data in zip(self.cache, next(self.child_epoch_iterator)):
             cache.extend(data)
 
 
+class NGramStream(CachedDataStream):
+    """Return n-grams from a stream.
+
+    This data stream wrapper takes as an input a data stream outputting
+    batches of sentences. From these sentences n-grams of a fixed order
+    (e.g. bigrams, trigrams, etc.) are extracted and returned.
+
```

Comment:
I think you should mention here that you are adding an additional source, and what this source is.